### PR TITLE
Implement Robust Cost Function

### PR DESCRIPTION
**Pull request by luisenp**
This replaces #138. It reuses some of the code in that PR (and all of the math implemented there) but reorganizes the classes differently. I went back to the version that had `RobustCostFunction` as a separate class and `Loss` as a basic functional form.

### CODE REVIEW

**Comment by luisenp** [comment]
This should work after #205 is merged.

File: theseus/core/robust_cost_function.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+import abc
```
Comment (mhmukadam) [inline theseus/core/robust_cost_function.py:1]: Missing license header

File: theseus/__init__.py [hunk 2]
```diff
@@ -7,11 +7,15 @@
     CostFunction,
     CostWeight,
     DiagonalCostWeight,
-    AutoDiffCostFunction,
     Objective,
     ScaleCostWeight,
     Variable,
     Vectorize,
+    Loss,
```
Comment (mhmukadam) [inline theseus/__init__.py:14]: Have we settled on Loss or are there any better alternatives? Loss good lead to some confusion of loss in the ML context.

File: theseus/core/loss.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import abc
+
+import torch
+
+
+class Loss:
```
Comment (mhmukadam) [inline theseus/core/loss.py:11]: We should add some unit tests for loss and robust loss.

Comment (mhmukadam) [reply theseus/core/loss.py:11]: This is an abstract class. Need to do `class Loss(abc.ABC):`

File: theseus/core/loss.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    @staticmethod
+    @abc.abstractmethod
+    def linearize(x: torch.Tensor, raidus: torch.Tensor) -> torch.Tensor:
+        pass
+
+
+class WelschLoss(Loss):
+    @staticmethod
+    def evaluate(x: torch.Tensor, radius: torch.Tensor) -> torch.Tensor:
```
Comment (mhmukadam) [inline theseus/core/loss.py:25]: Are there restrictions on the sizes of x and/or radius? Should we explicitly check for batch? If these are common checks then we can move it to the base class.

File: theseus/core/loss.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        return radius - radius * torch.exp(-x / radius)
```
Comment (mhmukadam) [inline theseus/core/loss.py:26]: Not sure if this is needed but do we raise a warning if radius is zero/close to zero. Incase you are learning it, there is a chance this fails silently by sending nan/infs.

File: theseus/core/robust_cost_function.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+# (see Theory section at http://ceres-solver.org/nnls_modeling.html#theory and
+# references therein); here we use alpha=0. The implementation of this part is done via
+# `RobustCostFunction.weighted_jacobians_error()`
+#
+# Let e := `robust_cost_fn.cost_function.error()`. Currently, the convention is that:
+#     -`robust_cost_fn.error()` returns e.
+#     -`robust_cost_fn.weighted_error()` returns a vectorized version of loss(||e||^2).
+#
+# Also, `robust_cost_fn.jacobians()` is not implemented.
+class RobustCostFunction(CostFunction, abc.ABC):
```
Comment (mhmukadam) [inline theseus/core/robust_cost_function.py:28]: I don't think this is an abstract class so we can remove `abc.ABC`.

File: theseus/core/robust_cost_function.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        # ||weighted_error||^2 = error_loss
+        # By doing this we avoid having to change the objective's error computation
+        # specifically for robust cost functions. The issue for this type of cost
+        # function is that the theory requires us to maintain scaled errors/jacobians
+        # of dim = robust_fn.cost_function.dim() to do the linearization properly,
+        # but the actual error has dim = 1, being the result of loss(||error||^2).
+        #
+        # Other options explored so far involve adding new methods to CostFunction,
+        # and/or changing Objective/Optimizer class. I'd prefer to avoid changing
+        # core class for experimental code, as long as it is possible [lep].
```
Comment (mhmukadam) [inline theseus/core/robust_cost_function.py:72]: Are these comments still needed (no longer experimental code)? Maybe needs some revision at least.

File: theseus/core/robust_cost_function.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        return (
+            torch.ones_like(weighted_error) * (error_loss / self.dim() + 1e-20).sqrt()
+        )
+
+    def jacobians(self) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        raise NotImplementedError
```
Comment (mhmukadam) [inline theseus/core/robust_cost_function.py:78]: Is the unweighted Jacobian not need anywhere? Maybe add a comment.

File: theseus/core/robust_cost_function.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        # specifically for robust cost functions. The issue for this type of cost
+        # function is that the theory requires us to maintain scaled errors/jacobians
+        # of dim = robust_fn.cost_function.dim() to do the linearization properly,
+        # but the actual error has dim = 1, being the result of loss(||error||^2).
+        #
+        # Other options explored so far involve adding new methods to CostFunction,
+        # and/or changing Objective/Optimizer class. I'd prefer to avoid changing
+        # core class for experimental code, as long as it is possible [lep].
+        return (
+            torch.ones_like(weighted_error) * (error_loss / self.dim() + 1e-20).sqrt()
```
Comment (mhmukadam) [inline theseus/core/robust_cost_function.py:74]: Hardcoded number, change to some EPS?

File: theseus/core/robust_cost_function.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        for attr in cost_function._aux_vars_attr_names:
+            setattr(self, attr, getattr(cost_function, attr))
+            self.register_aux_var(attr)
+
+        self.log_loss_radius = log_loss_radius
+        self.register_aux_var("log_loss_radius")
+        self.loss = loss_cls()
+
+    def error(self) -> torch.Tensor:
+        return self.cost_function.error()
```
Comment (mhmukadam) [inline theseus/core/robust_cost_function.py:54]: Should this return the robustified error (just not weighted)?
